src/project.js

```diff
@@ -0,0 +1,705 @@
+  setPaths (projectPaths, options = {}) {
+    for (let repository of this.repositories) {
+      if (repository != null) repository.destroy()
+    }
+    this.rootDirectories = []
+    this.repositories = []
+
+    for (let path in this.watcherPromisesByPath) {
+      this.watcherPromisesByPath[path].then(watcher => { watcher.dispose() })
+    }
+    this.watcherPromisesByPath = {}
+
+    const missingProjectPaths = []
+    for (let projectPath of projectPaths) {
+      try {
+        this.addPath(projectPath, {emitEvent: false, mustExist: true, exact: options.exact === true})
+      } catch (e) {
+        if (e.missingProjectPaths != null) {
+          missingProjectPaths.push(...e.missingProjectPaths)
+        } else {
+          throw e
+        }
+      }
+    }
+
+    this.emitter.emit('did-change-paths', projectPaths)
+
+    if ((options.mustExist === true) && (missingProjectPaths.length > 0)) {
+      const err = new Error('One or more project directories do not exist')
+      err.missingProjectPaths = missingProjectPaths
+      throw err
+    }
+  }
+
+  // Public: Add a path to the project's list of root paths
+  //
+  // * `projectPath` {String} The path to the directory to add.
+  // * `options` An optional {Object} that may contain the following keys:
+  //   * `mustExist` If `true`, throw an Error if the `projectPath` does not exist. If `false`, a `projectPath` that does
+  //     not exist is ignored. Default: `false`.
+  //   * `exact` If `true`, only add `projectPath` if it names an existing directory. If `false`, if `projectPath` is a
+  //     a file or does not exist, its parent directory will be added instead.
+  addPath (projectPath, options = {}) {
+    const directory = this.getDirectoryForProjectPath(projectPath)
+
+    let ok = true
+    if (options.exact === true) {
+      ok = (directory.getPath() === projectPath)
+    }
+    ok = ok && directory.existsSync()
+
+    if (!ok) {
+      if (options.mustExist === true) {
+        const err = new Error(`Project directory ${directory} does not exist`)
+        err.missingProjectPaths = [projectPath]
+        throw err
+      } else {
+        return
+      }
+    }
+
+    for (let existingDirectory of this.getDirectories()) {
+      if (existingDirectory.getPath() === directory.getPath()) { return }
+    }
+
+    this.rootDirectories.push(directory)
+    this.watcherPromisesByPath[directory.getPath()] = watchPath(directory.getPath(), {}, events => {
+      // Stop event delivery immediately on removal of a rootDirectory, even if its watcher
+      // promise has yet to resolve at the time of removal
+      if (this.rootDirectories.includes(directory)) {
+        this.emitter.emit('did-change-files', events)
+      }
+    })
+
+    for (let path in this.watcherPromisesByPath) {
+      if (!this.rootDirectories.find(dir => dir.getPath() === path)) {
+        this.watcherPromisesByPath[path].then(watcher => { watcher.dispose() })
+      }
+    }
```

Comment:
I think there's a bug lurking in the current CoffeeScript version of this `for` loop:

https://github.com/atom/atom/blob/8952cd315d44b7730ed686c2be14a9dc32d1e529/src/project.coffee#L288-L290

`@rootDirectories` is an Array of Directory objects, and `root` is a String. Therefore, `@rootDirectories.includes(root)` will always evaluate to `false`. We instead need to look for a Directory object in `@rootDirectories` where the Directory object's path is equal `root`.

The JavaScript version above (implemented in 5ec9d0f1347eda1aecea72ab00348a56eb3d0ec6) fixes this issue. (Note that this PR also renames the `root` local to `path` to clarify that we know the item is a _path_, but we don't know whether it is a _root_ path.)

Reply:
Oh yikes. I remember getting a bit confused about which collections contained paths and which contained Directory objects... looks like I missed one!

> Note that this PR also renames the `root` local to `path` to clarify that we know the item is a _path_, but we don't know whether it is a _root_ path.)

I'm :+1: on renaming `root` to clarify this, but I'd advise against calling it `path`. I've hit too many problems with inadvertently shadowing or overwriting the import from node's "path" module :smile: How about `rootPath`?

Reply:
@smashwilson: Thanks for the :eyes:

> I'm :+1: on renaming `root` to clarify this, but I'd advise against calling it `path`. I've hit too many problems with inadvertently shadowing or overwriting the import from node's "path" module :smile: 

Good point! a4ea46c57e078ae691d9aa9781948c39253ed4c9 renames it to `watchedPath`. (I chose `watchedPath` over `rootPath` because we know the path is watched, but we don't know if it's a root path.)